panels/updates/cc-updates-panel.c

```diff
@@ -0,0 +1,714 @@
+#include <NetworkManager.h>
+
+#define NM_SETTING_ALLOW_DOWNLOADS_WHEN_METERED "connection.allow-downloads-when-metered"
+#define NM_SETTING_TARIFF_ENABLED               "connection.tariff-enabled"
+#define NM_SETTING_TARIFF                       "connection.tariff"
+
+struct _CcUpdatesPanel
+{
+  CcPanel             parent;
+
+  GtkWidget          *automatic_updates_container;
+  GtkWidget          *automatic_updates_switch;
+  GtkWidget          *metered_data_label;
+  GtkWidget          *network_name_label;
+  GtkWidget          *network_settings_label;
+  GtkWidget          *network_status_icon;
+  GtkWidget          *scheduled_updates_container;
+  GtkWidget          *scheduled_updates_switch;
+  CcTariffEditor     *tariff_editor;
+
+  /* Network Manager */
+  NMClient           *nm_client;
+  NMDevice           *current_device;
+
+  /* Signal handlers */
+  gulong              changed_id;
+  guint               save_tariff_timeout_id;
+
+  GCancellable       *cancellable;
+};
+
+
+static void          on_automatic_updates_switch_changed_cb      (GtkSwitch      *sw,
+                                                                  GParamSpec     *pspec,
+                                                                  CcUpdatesPanel *self);
+
+static void          on_network_changed_cb                       (CcUpdatesPanel *self);
+
+static void          on_network_changes_committed_cb             (GObject        *source,
+                                                                  GAsyncResult   *result,
+                                                                  gpointer        user_data);
+
+static void          on_scheduled_updates_switch_changed_cb      (GtkSwitch      *sw,
+                                                                  GParamSpec     *pspec,
+                                                                  CcUpdatesPanel *self);
+
+static void          on_tariff_changed_cb                        (CcTariffEditor *tariff_editor,
+                                                                  CcUpdatesPanel *self);
+
+
+G_DEFINE_TYPE (CcUpdatesPanel, cc_updates_panel, CC_TYPE_PANEL)
+
+enum
+{
+  PROP_PARAMETERS = 1,
+  N_PROPS
+};
+
+
+/*
+ * Auxiliary methods
+ */
+
+static void
+cleanup_signals (CcUpdatesPanel *self)
+{
+  if (self->changed_id > 0)
+    {
+      g_signal_handler_disconnect (self->current_device, self->changed_id);
+      self->changed_id = 0;
+    }
+}
+
+static void
+store_automatic_updates_setting (NMConnection *connection,
+                                 gboolean      automatic_updates_enabled,
+                                 gboolean      tariff_enabled,
+                                 GVariant     *tariff_variant)
+{
+  NMSettingUser *setting_user;
+  g_autofree gchar *tariff_string = NULL;
+  g_autoptr(GError) error = NULL;
+
+  setting_user = NM_SETTING_USER (nm_connection_get_setting (connection, NM_TYPE_SETTING_USER));
+  g_assert_nonnull (setting_user);
```

Comment:
Nitpick: `g_assert_nonnull()` should only be used for unit tests, since it’s not compiled out when compiling with `G_DISABLE_CHECKS`. `g_assert()` is (and hence you get improved performance with `G_DISABLE_CHECKS`). Use `g_assert()` in application/library code, and `g_assert_*()` in unit tests.